File: python/string/17. Letter Combinations of a Phone Number.py
'''
BackTracking:
Time: O(4^n)  -- 4: 'pqrs' or 'wxyz'
Space: O(4^n)
'''

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        res = []
        keyboard = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz"
        }

        def backTrack(i, curStr):
            if len(curStr) == len(digits):
                res.append(curStr)
                return
            for c in keyboard[digits[i]]:  # each recursive call: 3~4 iters
                backTrack(i + 1, curStr + c)

        if digits:
            backTrack(0, "")

        return res

File: javascript/string/17. Letter Combinations of a Phone Number.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function(digits) {
    const res = [];
    const keyboard = {
        "2": "abc",
        "3": "def",
        "4": "ghi",
        "5": "jkl",
        "6": "mno",
        "7": "pqrs",
        "8": "tuv",
        "9": "wxyz"
    };

    var backTrack = (i, curStr) => {
        if (digits.length == curStr.length) {
            res.push(curStr);
            return;
        }
        for (let c of keyboard[digits[i]]) backTrack(i + 1, curStr + c);
    }
    if (digits) backTrack(0, "");

    return res;   
};